org.eclipse.m2e.pde/src/org/eclipse/m2e/pde/CacheManager.java

```diff
@@ -0,0 +1,127 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Christoph Läubrich
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *      Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.m2e.pde;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.io.FileUtils;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.pde.core.target.ITargetHandle;
+import org.eclipse.pde.core.target.TargetBundle;
+
+public class CacheManager {
+
+	private static final String LASTACCESS_MARKER = ".lastaccess";
+
+	private static File baseDir;
+
+	private static final Map<String, CacheManager> MANAGERS = new HashMap<>();
+
+	private volatile boolean invalidated;
+
+	private File folder;
+
+	public CacheManager(File folder) {
+		this.folder = folder;
+		try {
+			FileUtils.touch(new File(LASTACCESS_MARKER));
+		} catch (IOException e) {
+			// can't mark last access then...
+		}
+	}
+
+	public synchronized <R> R accessArtifactFile(Artifact artifact, CacheConsumer<R> consumer) throws Exception {
```

Comment:
Why not just `Funtion<File, R>` instead of CacheConsumer here?

Reply:
Functions are not allowed to throw (checked) exceptions.